crates/accelerate/src/sparse_pauli_op.rs

```diff
@@ -337,163 +338,463 @@ pub fn decompose_dense(
     tolerance: f64,
 ) -> PyResult<ZXPaulis> {
     let num_qubits = operator.shape()[0].ilog2() as usize;
-    let size = 1 << num_qubits;
-    if operator.shape() != [size, size] {
+    let side = 1 << num_qubits;
+    if operator.shape() != [side, side] {
         return Err(PyValueError::new_err(format!(
-            "input with shape {:?} cannot be interpreted as a multiqubit operator",
+            "bad input shape for {} qubits: {:?}",
+            num_qubits,
             operator.shape()
         )));
     }
-    let mut paulis = vec![];
-    let mut coeffs = vec![];
-    if num_qubits > 0 {
-        decompose_dense_inner(
-            C_ONE,
-            num_qubits,
-            &[],
-            operator.as_array(),
-            &mut paulis,
-            &mut coeffs,
-            tolerance * tolerance,
-        );
-    }
-    if coeffs.is_empty() {
-        Ok(ZXPaulis {
-            z: PyArray2::zeros_bound(py, [0, num_qubits], false).into(),
-            x: PyArray2::zeros_bound(py, [0, num_qubits], false).into(),
-            phases: PyArray1::zeros_bound(py, [0], false).into(),
-            coeffs: PyArray1::zeros_bound(py, [0], false).into(),
-        })
-    } else {
-        // Constructing several arrays of different shapes at once is rather awkward in iterator
-        // logic, so we just loop manually.
-        let mut z = Array2::<bool>::uninit([paulis.len(), num_qubits]);
-        let mut x = Array2::<bool>::uninit([paulis.len(), num_qubits]);
-        let mut phases = Array1::<u8>::uninit(paulis.len());
-        for (i, paulis) in paulis.drain(..).enumerate() {
-            let mut phase = 0u8;
-            for (j, pauli) in paulis.into_iter().rev().enumerate() {
-                match pauli {
-                    Pauli::I => {
-                        z[[i, j]].write(false);
-                        x[[i, j]].write(false);
-                    }
-                    Pauli::X => {
-                        z[[i, j]].write(false);
-                        x[[i, j]].write(true);
-                    }
-                    Pauli::Y => {
-                        z[[i, j]].write(true);
-                        x[[i, j]].write(true);
-                        phase = phase.wrapping_add(1);
-                    }
-                    Pauli::Z => {
-                        z[[i, j]].write(true);
-                        x[[i, j]].write(false);
-                    }
+    let (stack, mut out_list, mut scratch) = decompose_first_level(operator.as_array(), num_qubits);
+    decompose_middle_levels(stack, &mut out_list, &mut scratch, num_qubits);
+    let out = decompose_last_level(&mut out_list, &scratch, num_qubits, tolerance)?;
```

Comment:
The original implementation of the code in Python is less than 100 lines, and looks much simpler.
It's worth to add some explanation on dividing this code into 3 levels, which is done in order to reduce the memory.